### PR TITLE
Implement `query packet pending` command

### PR DESCRIPTION
**Pull request by mzabaluev**
Closes: #1862

## Description

Add a new subcommand `pending` to `query packet`, displaying a summary of pending packet commitment IDs for the identified channel and its reverse counterpart.

______

### PR author checklist:
- [x] Added changelog entry, using [`unclog`](https://github.com/informalsystems/unclog).
- [x] Added tests: integration (for Hermes) or unit/mock tests (for modules). 
- [x] Linked to GitHub issue.
- [ ] Updated code comments and documentation (e.g., `docs/`).

### Reviewer checklist:

- [ ] Reviewed `Files changed` in the GitHub PR explorer.
- [ ] Manually tested (in case integration/unit/mock tests are absent).

### CODE REVIEW

**Comment by mzabaluev** [comment]
The command name could be subject to bikeshedding. It does not feel like a proper member of the `query packet` family.
Can it be `query packets` or `list packets`? The rationale is, it's a high level command displaying a summary of all packet commitments currently known as being in some pending state on both ends of the channel.

File: tools/integration-test/src/tests/query_packet.rs
Comment (mzabaluev) [inline tools/integration-test/src/tests/query_packet.rs:73]: @soareschen Should there be an easier way to instantiate an `IdentifiedChannelEnd`, maybe as a method of `ConnectedChannel`?

Comment (soareschen) [reply tools/integration-test/src/tests/query_packet.rs:73]: Yes, I think we can probably add a method in `ibc_test_framework::relayer::channel::TaggedChannelEndExt`. 

```rust
trait TaggedChannelEndExt<ChainA, ChainB> {
    fn as_identified_channel_end(&self) -> DualTagged<ChainA, ChainB, IdentifiedChannelEnd>;
}
```

Otherwise there is a tagged version of `query_channel` at `ibc_test_framework::relayer::channel::query_channel_end`. We could alternatively add a new function like `query_identified_channel_end` there.

```rust
fn query_identified_channel_end<ChainA: ChainHandle, ChainB>(
    handle: &ChainA,
    channel_id: &TaggedChannelIdRef<ChainA, ChainB>,
    port_id: &TaggedPortIdRef<ChainA, ChainB>,
) -> Result<DualTagged<ChainA, ChainB, IdentifiedChannelEnd>, Error>
```

The tagged version of query channel helps us ensure that we do not mix up the channel identifiers some where along the way.

File: tools/integration-test/src/tests/query_packet.rs
Comment (seanchen1991) [inline tools/integration-test/src/tests/query_packet.rs:17]: Just want to clarify: for the purposes of this test, is it fine that pending packets are relayed once the relayer has been started? That's what will happen if `config.mode.tx_confirmation` is `true`, but I don't know what the default value of `tx_confirmation` is. 

If it's `false`, then pending packets won't be cleared due to the interaction between `tx_confirmation` and `clear_interval`. If it's `true`, then they will be cleared on account of the interaction. 

If it doesn't matter that pending packets are cleared after the relayer has been started, then I think it's also fine to ignore this interaction.

File: tools/integration-test/src/tests/query_packet.rs
Comment (seanchen1991) [inline tools/integration-test/src/tests/query_packet.rs:29]: We can remove this method since the trait implementation defaults to `Order::Unordered`.

File: tools/integration-test/src/tests/query_packet.rs
Comment (seanchen1991) [inline tools/integration-test/src/tests/query_packet.rs:63]: I've seen in other integration tests that the process sleeps for 1 second. Is there any particular rationale for having it sleep for 2 in this case?

File: tools/integration-test/src/tests/query_packet.rs
Comment (seanchen1991) [inline tools/integration-test/src/tests/query_packet.rs:124]: Same question here about sleeping for 2 seconds vs 1.

**Review by seanchen1991 — DISMISSED** [review]
Left a few comments and questions. There are few places where the variable name `ccc` is used. Could we replace these with more descriptive variable names?

File: tools/integration-test/src/tests/query_packet.rs
Comment (mzabaluev) [reply tools/integration-test/src/tests/query_packet.rs:63]: I had one sporadically failed run with 1 second sleep here, and none after I bumped it to 2.
It would be good to have some way to wait for confirmation.

File: relayer-cli/src/commands/query/packet/acks.rs
Comment (mzabaluev) [reply relayer-cli/src/commands/query/packet/acks.rs:39]: Yeah... The type is a carry bag for returning channel-connection-client, it's never used as a parameter, so it's always clear from use of the members what are we accessing. Will rename for further clarity.

File: relayer-cli/src/commands/query/packet/pending.rs
Comment (mzabaluev) [reply relayer-cli/src/commands/query/packet/pending.rs:19]: It's important to bikeshed this indeed, because these names show up in serialization and output.
No objections to `src` and `dst` from me.

File: tools/integration-test/src/tests/query_packet.rs
Comment (mzabaluev) [reply tools/integration-test/src/tests/query_packet.rs:17]: The test never runs the supervisor, packets are relayed step by step by invoking methods on a `Link`.
This is to get the packet summaries into an expected state without dependency on timing.

**Comment by mzabaluev** [comment]
The test failure looks like an unrelated flaky test.

**Comment by ancazamfir** [comment]
Ran a few tests, things look good. A bit hard to read with lots of packets. Have you tried to implement the collated output?

**Comment by ancazamfir** [comment]
So the output looks something like this:
```
Success: Summary {
    forward: PendingPackets {
        unreceived: [
            2203,
...
            2212,
        ],
        pending_acks: [
            2183,
...
           2202,
        ],
    },
    reverse: PendingPackets {
        unreceived: [
            14,
...
           23,
        ],
        pending_acks: [
            4,
...
            13,
        ],
    },
}
```
Looking at the field names, since all are unreceived/ pending, maybe `unreceived` -> `pending_recv`?
So we have `pending_recv` and `pending_acks`?
Or have `unreceived_packets` and `unreceived_acks`? Though this looks weird as sometimes we use `packets` to refer to all types (i.e. send, recv, timeout, ack)

**Comment by mzabaluev** [comment]
> Have you tried to implement the collated output?

Not yet. I can do it, or maybe leave this for a later enhancement? The other `query packet` subcommands could benefit from this as well.

**Comment by romac** [comment]
> Looking at the field names, since all are unreceived/ pending, maybe `unreceived` -> `pending_recv`? So we have `pending_recv` and `pending_acks`?

+1 for `pending_recv` and `pending_acks`

**Comment by romac** [comment]
> IMO we can leave it like this for this PR and clean things up in a subsequent PR.

+1

**Comment by romac** [comment]
@mzabaluev is this ready for merge or do you have more pending changes? are we leaving the collated output for a later PR?

**Comment by mzabaluev** [comment]
@romac We can leave the collation for later; I've filed it as #2142.

**Comment by romac** [comment]
Great, then I'll leave it to @seanchen1991 to take another look at this and merge.

**Review by seanchen1991 — COMMENTED** [review]
It looks like there are still a few lingering clippy lints reporting unnecessary `clone` calls. Are these calls being left in intentionally?

**Comment by adizere** [comment]
> It looks like there are still a few lingering clippy lints reporting unnecessary `clone` calls. Are these calls being left in intentionally?

Good point, we should fix those!

**Review by romac — APPROVED** [review]

**Pull request merged**
